flax/nnx/bridge/module.py

```diff
@@ -0,0 +1,493 @@
+      for val, abs_val in zip(value_flat, abs_value_flat):
+        if jnp.shape(val) != jnp.shape(abs_val):
+          raise errors.ScopeParamShapeError(
+            name, '', jnp.shape(abs_val), jnp.shape(val)
+          )
+
+      if isinstance(abs_value, variablelib.VariableMetadata):
+        abs_value.raw_value = value
+        value = abs_value
+
+      variable = variable_type(value, unbox=unbox)
+    else:
+      if init_fn is None:
+        raise ValueError(f"Expected 'init_fn' to be a callable, got None")
+
+      value = init_fn(*init_args, **init_kwargs)
+      variable = variable_type(value, unbox=unbox)
+
+    setattr(self, name, variable)
+    return variable
+
+  def _get_variables(self):
+    state = graph.state(self)
+    _variables: tp.Mapping = {}
+
+    variable_state: variablelib.VariableState
+    for path, variable_state in state.flat_state():
+      try:
+        collection = variablelib.variable_name_from_type(variable_state.type)
+      except ValueError:
+        collection = variable_state.type.__name__
+
+      if collection not in _variables:
+        _variables[collection] = {}
+
+      if has_unbox(variable_state):
+        assert isinstance(variable_state, variablelib.VariableState)
+        if variable_state.unbox:
+          leaf = variable_state.value
+        else:
+          # TODO: convert to linen box
+          leaf = variable_state
+      else:
+        leaf = variable_state.value
+
+      _variables[collection][path] = leaf
+
+    _variables = {
+      collection: traversals.unflatten_mapping(flat_state)
+      for collection, flat_state in _variables.items()
+    }
+
+    return _variables
+
+  @property
+  def variables(self):
+    _variables = FrozenDict(self._get_variables())
+    return _variables
+
+  def apply(
+    self,
+    variables: dict[str, tp.Mapping],
+    *args,
+    rngs: int | jax.Array | dict[str, jax.Array] | rnglib.Rngs | None = None,
+    method: tp.Callable[..., tp.Any] | str = '__call__',
+    mutable: tp.Any = False,
+    _initialize: bool = False,
+    **kwargs,
+  ) -> tp.Any:
+    module = graph.clone(self)
+
+    # create variables
+    real_variables = dict(variables)
+    for collection, state in variables.items():
+      variable_type = variablelib.variable_type_from_name(
+        collection, allow_register=True
+      )
+
+      def to_variable(value):
+        if isinstance(value, variablelib.VariableState):
+          variable = value.to_variable()
+          if not isinstance(variable, variable_type):
+            raise ValueError(
+              f'Expected variable to be of type {variable_type}, got {type(variable).__name__}'
+            )
+          return variable
+        return variable_type(value)
+
+      state = jax.tree.map(to_variable, state)
+      real_variables[collection] = state
+
+    states = ({},) if not real_variables else real_variables.values()
+    graph.update(module, *states)
+
+    if rngs is None:
+      rngs = rnglib.Rngs()
```

Comment:
Is this instance able to yield anything? If not would it be less confusing to use `None`?

Reply:
This is not able to yield anything, you'll get an error but I think its fine as this is what happens in Linen as well. We could also set it to `None` and then raise the error manually. For now relying on `Rngs` to do the error handling is a bit simpler.